[lldb/Commands] Add `scripting template list` command with auto discovery

medismailben opened the pull request:
This patch introduces a new `template` multiword sub-command to the
`scripting` top-level command. As the name suggests, this sub-command
operates on scripting templates, and currently has the ability to
automatically discover the various scripting extensions that lldb
supports.

Signed-off-by: Med Ismail Bennani <ismail@bennani.ma>

medismailben commented:
Depends on #97263

bulbazord commented on lldb/source/Commands/CommandObjectScripting.cpp:
```diff
@@ -127,6 +130,171 @@ class CommandObjectScriptingRun : public CommandObjectRaw {
+    s.Printf("Available scripted affordances:\n");
+
+    auto print_field = [&s](llvm::StringRef key, llvm::StringRef value,
+                            bool check_validy = false) {
```
typo: `validy` -> `validity`?


Also, what's the use case for the check? I don't see you use it.

medismailben replied:
It's used in the next line. If set, it will check that the value argument isn't empty.

bulbazord commented on lldb/source/Commands/CommandObjectScripting.cpp:
```diff
@@ -127,6 +130,171 @@ class CommandObjectScriptingRun : public CommandObjectRaw {
+    }
+
+    Stream &s = result.GetOutputStream();
+    s.Printf("Available scripted affordances:\n");
```
Suggestion: `Available scripting templates`. My reason is that the command is `scripting template list` so I would expect it to tell me what the available scripting templates are.

Additionally, it might be useful to specify the language. So something like `Available scripting templates (Python)` or something like that. This would let you drop the `print_field("Language", "Python")` below.

medismailben replied:
I was thinking `scripting extensions` rather than `scripting templates` may be we can do both `scripting extension templates` here ?

For the second suggestion, the command has an optional `-l|--language  <script-language>` option to only print extension information for that specific language.

bulbazord commented on lldb/source/Commands/CommandObjectScripting.cpp:
```diff
@@ -127,6 +130,171 @@ class CommandObjectScriptingRun : public CommandObjectRaw {
+    size_t i = 0;
+    for (llvm::StringRef plugin_name =
+             PluginManager::GetScriptedInterfaceNameAtIndex(i);
+         !plugin_name.empty();) {
```
Increment `i` here instead of below? It would be more resilient to making mistakes when modifying this loop for any reason.

bulbazord commented on lldb/source/Plugins/CMakeLists.txt:
```diff
@@ -29,6 +29,9 @@ add_subdirectory(UnwindAssembly)
 set(LLDB_STRIPPED_PLUGINS)
 get_property(LLDB_ALL_PLUGINS GLOBAL PROPERTY LLDB_PLUGINS)
+get_property(LLDB_EXTRA_PLUGINS GLOBAL PROPERTY LLDB_EXTRA_SCRIPT_PLUGINS)
+list(APPEND LLDB_ALL_PLUGINS ${LLDB_EXTRA_PLUGINS})
```
I don't think there's a benefit to splitting the plugins into two different global properties. Is there a reason why we should treat these script plugins differently? 

Also I see you doing that for the python plugins below, how about the Lua ones?

medismailben replied:
I tried reusing `LLDB_ALL_PLUGINS` in `lldb/source/Plugins/ScriptInterpreter/Python/Interfaces/CMakeLists.txt` but I couldn't make it work: Either the scripted interface plugins wouldn't register or we would fail to link.

Not all the Python interfaces are registered so far, same for Lua, they will be added over time.

I'll leave it as is for now but feel free to change this after it lands :)

JDevlieghere replied:
What exactly didn't work? I can't assess if this workaround makes sense without understanding the problem. It's totally fine for some of those interfaces not to be plugins.

bulbazord replied:
I see why this doesn't work now. LLDB plugins are added to the `LLDB_PLUGINS` global property via the PLUGIN argument to `add_lldb_library`. You aren't adding the PLUGIN argument to the `add_lldb_library` invocation in `lldb/source/Plugins/ScriptInterpreter/Python/Interfaces/CMakeLists.txt`. This is how the build system knows the list of plugins.

But even if you were to remove this and add that `PLUGIN` argument, this still wouldn't work. I don't think we've explicitly documented this behavior, but LLDB expects each library to be one plugin, not three as you've done here. When LLDB creates a plugin, it's going to generate a `Plugins.def` file that LLDB uses to initialize and teardown all plugins. It has a list of entries that look like this:
`LLDB_PLUGIN(lldbPluginFooBar)`. There are some exceptions for the script interpreter plugins (not sure why, didn't check). It also expects the `LLDB_PLUGIN_DEFINE` argument to match the library name exactly (without the initial `lldbPlugin`, specifically).

The way I see it, you have three options:
1. Restructure the build so that there are 3 libraries, each one of them a plugin.
2. Change LLDB's CMake machinery to accept libraries that define multiple plugin interfaces. You'll need to modify `add_lldb_library` so that you can pass it a list of plugin interface names instead of assuming the plugin names matches the library's name.
3. Keep the workaround and document why it exists.

I would suggest (1) or (2). I think (3) is just kicking the can down the road. This is a solution that explicitly works around LLDB's existing plugin system, is difficult to discover without knowing its there, and could make future refactors more difficult.

jimingham commented on lldb/source/Commands/CommandObjectScripting.cpp:
```diff
@@ -127,6 +130,171 @@ class CommandObjectScriptingRun : public CommandObjectRaw {
+      return llvm::ArrayRef(g_scripting_template_list_options);
+    }
+
+    lldb::ScriptLanguage language = lldb::eScriptLanguageNone;
```
should be m_language, right?

medismailben replied:
Correct

medismailben commented:
FWIW, this is the output it's producing for now:
```
(lldb) scripting template list 
Available scripted affordances:
  Name: ScriptedProcessPythonInterface
  Language: Python
  Description: Mock process state
  Command Interpreter Usages:
    process attach -C <script-name> [-k key -v value ...]
    process launch -C <script-name> [-k key -v value ...]
  API Usages:
    SBAttachInfo.SetScriptedProcessClassName
    SBAttachInfo.SetScriptedProcessDictionary
    SBTarget.Attach
    SBLaunchInfo.SetScriptedProcessClassName
    SBLaunchInfo.SetScriptedProcessDictionary
    SBTarget.Launch

  Name: ScriptedThreadPlanPythonInterface
  Language: Python
  Description: Alter thread stepping logic and stop reason
  Command Interpreter Usages: thread step-scripted -C <script-name> [-k key -v value ...]
  API Usages: SBThread.StepUsingScriptedThreadPlan
```

jimingham commented on lldb/source/Commands/CommandObjectScripting.cpp:
```diff
@@ -127,6 +130,171 @@ class CommandObjectScriptingRun : public CommandObjectRaw {
+            ? m_interpreter.GetDebugger().GetScriptLanguage()
+            : m_options.language;
+
+    if (language == lldb::eScriptLanguageNone) {
```
Given that the vast majority of our users use Python as their scripting language, could we default to that rather than None?

jimingham replied:
Or maybe better eScriptLanguageDefault?

jimingham commented on lldb/source/Commands/CommandObjectScripting.cpp:
```diff
@@ -127,6 +130,171 @@ class CommandObjectScriptingRun : public CommandObjectRaw {
+      print_field("Name", plugin_name);
+      switch (lang) {
+      case eScriptLanguagePython:
+        print_field("Language", "Python");
```
Use ScriptInterpreter::LanguageToString rather than hard coding the values here.

jimingham commented on lldb/source/Commands/Options.td:
```diff
@@ -841,6 +841,12 @@ let Command = "scripting run" in {
+let Command = "scripting template list" in {
+  def scripting_template_list_language : Option<"language", "l">,
+    EnumArg<"ScriptLang">, Desc<"Specify the scripting "
+    " language. If none is specific the default scripting language is used.">;
```
specific -> specified

jimingham replied:
I don't think that's what is currently implemented, however, since you reset the language to None, and then error out if it is...

jimingham reviewed (approved):
LGTM

JDevlieghere commented on lldb/source/Commands/CommandObjectScripting.cpp:
```diff
@@ -127,6 +130,157 @@ class CommandObjectScriptingRun : public CommandObjectRaw {
+  CommandObjectScriptingTemplateList(CommandInterpreter &interpreter)
+      : CommandObjectParsed(
+            interpreter, "scripting template list",
+            "List all the available scripting affordances templates. ",
```
`affordances templates/affordance templates`

JDevlieghere commented on lldb/include/lldb/Core/PluginManager.h:
```diff
@@ -487,6 +487,33 @@ class PluginManager {
+  RegisterPlugin(llvm::StringRef name, llvm::StringRef description,
+                 ScriptedInterfaceCreateInstance create_callback,
+                 lldb::ScriptLanguage language,
+                 std::vector<llvm::StringRef> command_interpreter_usages,
+                 std::vector<llvm::StringRef> api_usages);
```
Could the usages be their own objects? It would simplif the plugin API and if you add a dump method you could simplify the command object too.

JDevlieghere commented on lldb/include/lldb/Core/PluginManager.h:
```diff
@@ -487,6 +487,33 @@ class PluginManager {
+
+  static bool UnregisterPlugin(ScriptedInterfaceCreateInstance create_callback);
+
+  static ScriptedInterfaceCreateInstance
+  GetScriptedInterfaceCreateCallbackAtIndex(uint32_t idx);
```
This looks unused?

JDevlieghere commented on lldb/include/lldb/Interpreter/Interfaces/ScriptedInterface.h:
```diff
@@ -68,6 +68,13 @@ class ScriptedInterface {
     return true;
   }
 
+  static bool
+  CreateInstance(lldb::ScriptLanguage language,
+                 std::vector<llvm::StringRef> command_interpreter_usages,
+                 std::vector<llvm::StringRef> api_usages) {
+    return false;
+  }
```
Without `GetScriptedInterfaceCreateCallbackAtIndex` I don't think you need this either.

JDevlieghere commented on lldb/include/lldb/lldb-private-interfaces.h:
```diff
@@ -124,6 +124,10 @@ typedef lldb::REPLSP (*REPLCreateInstance)(Status &error,
                                            lldb::LanguageType language,
                                            Debugger *debugger, Target *target,
                                            const char *repl_options);
+typedef bool (*ScriptedInterfaceCreateInstance)(
+    lldb::ScriptLanguage language,
+    std::vector<llvm::StringRef> command_interpreter_usages,
+    std::vector<llvm::StringRef> api_usages);
```
Nor this.

JDevlieghere commented on lldb/source/Commands/CommandObjectScripting.cpp:
```diff
@@ -127,6 +130,157 @@ class CommandObjectScriptingRun : public CommandObjectRaw {
+      }
+    };
+
+    auto print_usages = [&s](llvm::StringRef usage_kind,
+                             std::vector<llvm::StringRef> &usages) {
+      s.IndentMore();
+      s.Indent();
+      s << usage_kind << " Usages:";
+      if (usages.empty())
+        s << " None\n";
+      else if (usages.size() == 1)
+        s << " " << usages.front() << '\n';
+      else {
+        s << '\n';
+        for (llvm::StringRef usage : usages) {
+          s.IndentMore();
+          s.Indent();
+          s << usage << '\n';
+          s.IndentLess();
+        }
+      }
+      s.IndentLess();
+    };
```
This could be part of the class I was suggestion earlier.

JDevlieghere commented on lldb/source/Commands/CommandObjectScripting.cpp:
```diff
@@ -127,6 +130,157 @@ class CommandObjectScriptingRun : public CommandObjectRaw {
+
+    auto print_field = [&s](llvm::StringRef key, llvm::StringRef value,
+                            bool check_validity = false) {
+      if (!check_validity || !value.empty()) {
```
What's `check_validity`? It doesn't look like it's ever set.

JDevlieghere replied:
You marked this as resolved because you check its value, but that variable is still never set?

JDevlieghere commented on lldb/source/Commands/CommandObjectScripting.cpp:
```diff
@@ -127,6 +130,157 @@ class CommandObjectScriptingRun : public CommandObjectRaw {
+
+#pragma mark CommandObjectMultiwordScriptingTemplate
+
+// CommandObjectMultiwordScriptingTemplate
```
Remove

medismailben commented on lldb/source/Plugins/ScriptInterpreter/Python/Interfaces/CMakeLists.txt:
```diff
@@ -19,6 +19,12 @@ if (LLDB_ENABLE_LIBEDIT)
 endif()
 
+set_property(GLOBAL PROPERTY LLDB_EXTRA_SCRIPT_PLUGINS
+  lldbPluginOperatingSystemPythonInterface
+  lldbPluginScriptedProcessPythonInterface
+  lldbPluginScriptedThreadPlanPythonInterface
```
@bulbazord aren't we making 3 separate libraries here ?

bulbazord replied:
No, you're just adding some elements to the list `LLDB_EXTRA_SCRIPT_PLUGINS`. Right below this is `add_lldb_library` which creates one library. It will contain the 3 cpp files that implement these plugin interfaces.

bulbazord commented on lldb/include/lldb/Target/DynamicLoader.h:
```diff
@@ -329,6 +330,13 @@ class DynamicLoader : public PluginInterface {
+  /// Return the `start` function \b symbol in the dynamic loader module.
+  /// This is the symbol the process will begin executing with
+  /// `process launch --stop-at-entry`.
+  virtual std::optional<lldb_private::Symbol> GetStartSymbol() {
```
You'll probably need to rebase this change right?

bulbazord commented on lldb/source/Commands/CommandObjectScripting.cpp:
```diff
@@ -127,9 +129,127 @@ class CommandObjectScriptingRun : public CommandObjectRaw {
+            option_arg, GetDefinitions()[option_idx].enum_values,
+            eScriptLanguageNone, error);
+        if (!error.Success())
+          error.SetErrorStringWithFormat("unrecognized value for language '%s'",
```
Suggestion: `SetErrorStringWithFormatv` to avoid the `.str().c_str()` dance

bulbazord reviewed (commented):
Seems alright to me, left a few minor suggestions inline.

bulbazord reviewed (approved):
LGTM, you'll probably want sign off from Jonas too though since he had a ton of comments.

JDevlieghere reviewed (approved):
LGTM modulo comment.

medismailben committed:
[lldb/Commands] Add `scripting template list` command with auto discovery

This patch introduces a new `template` multiword sub-command to the
`scripting` top-level command. As the name suggests, this sub-command
operates on scripting templates, and currently has the ability to
automatically discover the various scripting extensions that lldb
supports.

Signed-off-by: Med Ismail Bennani <ismail@bennani.ma>

mysterymath commented:
The reland broke the Fuchsia Windows build with a similar error to the LLVM Windows bots, but I wasn't able to cleanly revert this commit without detailed knowledge of the contents. Could you revert this (preferentially) or fix forward?

medismailben commented:
> The reland broke the Fuchsia Windows build with a similar error to the LLVM Windows bots, but I wasn't able to cleanly revert this commit without detailed knowledge of the contents. Could you revert this (preferentially) or fix forward?
> 
> https://luci-milo.appspot.com/ui/p/fuchsia/builders/toolchain.ci/clang-windows-x64/b8741458989697976961/overview
> 
> https://logs.chromium.org/logs/fuchsia/buildbucket/cr-buildbucket/8741458989697976961/+/u/clang/build/stdout

Looking at this. Thanks

zeroomega commented:
Hi @medismailben 
We are seeing redefinition of pid_t error again on windows builder: https://lab.llvm.org/buildbot/#/builders/141/builds/1254
after multiple revert and reland like e72cdae47b4e263ea97b2bdd75cf44c1510cf3be. Could you revert the change please? Could you also not testing code in production please? It is a bad practice.

DavidSpickett commented:
I thought I saw that fixed once already, but perhaps that was just a full revert.

We (Linaro) can test a patch for you if it's fiddly to verify, we just need to know the general idea of it.

DavidSpickett commented:
Actually I'm looking at old results, the latest build makes it to tests https://lab.llvm.org/buildbot/#/builders/141/builds/1285 (which fail for reasons unrelated to this patch).

medismailben commented:
Hey @DavidSpickett! Thanks for the help! I don't have a system to try this on and it only fails on windows :/

> Actually I'm looking at old results, the latest build makes it to tests https://lab.llvm.org/buildbot/#/builders/141/builds/1285 (which fail for reasons unrelated to this patch).

My patch has already been reverted in 9effefbae8d96006a4dd29bb9ab8532fd408559d and the failure you're talking about was introduced in #100710. I believe @Michael137 is trying to fix it.

If you have a recent build of lldb on windows (with python enabled), could you run the `scripting template list` command and paste the output here.

I've split this patch in smaller chunks (bccff3baeff8, ecf125eaf0b1, 995f643f5aa7) and they've all landed without any build failure on windows except for 2402b3213c2f. I've noticed however that the cmake configure phase would produce some warnings on the bot regarding the files related to my changes, about the fact that they'd exceed `CMAKE_OBJECT_PATH_MAX` limit (250 on windows apparently).

You can see it happening here : https://lab.llvm.org/buildbot/#/builders/141/builds/1222

```
CMake Warning in C:/Users/tcwg/llvm-worker/lldb-aarch64-windows/llvm-project/lldb/source/Plugins/ScriptInterpreter/Python/Interfaces/ScriptedProcessPythonInterface/CMakeLists.txt:
  The object file directory
    C:/Users/tcwg/llvm-worker/lldb-aarch64-windows/build/tools/lldb/source/Plugins/ScriptInterpreter/Python/Interfaces/ScriptedProcessPythonInterface/CMakeFiles/lldbPluginScriptInterpreterPythonScriptedProcessPythonInterface.dir/./
  has 227 characters.  The maximum full path to an object file is 250
  characters (see CMAKE_OBJECT_PATH_MAX).  Object file
    ScriptedProcessPythonInterface.cpp.obj
  cannot be safely placed under this directory.  The build may not work
  correctly.
```

Also, if you could re-apply e72cdae47b4e263ea97b2bdd75cf44c1510cf3be and attach your build log here, that would be very helpful, and hopefully help us narrow down the ambiguous typedef issue. 

If I'm not able to make it work on windows, I'll unfortunately have to compile it out of lldb for that platform, since I don't have access to a machine to investigate the issue further.

Thank you in advance!

DavidSpickett commented:
I got it to build - https://github.com/llvm/llvm-project/pull/101672.
